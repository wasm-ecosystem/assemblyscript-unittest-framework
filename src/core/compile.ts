import { main } from "assemblyscript/asc";
import { join, relative } from "node:path";
import { findRoot } from "../utils/pathResolver.js";

export async function compile(testCodePaths: string[], outputFolder: string, compileFlags: string): Promise<string[]> {
  const wasm: string[] = [];
  const root = findRoot(testCodePaths);
  const compile = async (testCodePath: string) => {
    const outputWasm = getNewPath(outputFolder, root, testCodePath).slice(0, -2).concat("wasm");
    wasm.push(outputWasm);
    const outputWat = getNewPath(outputFolder, root, testCodePath).slice(0, -2).concat("wat");
    let ascArgv = [
      testCodePath,
      "--outFile",
      outputWasm,
      "--textFile",
      outputWat,
      "--exportStart",
      "_start",
      "--sourceMap",
      "--debug",
      "-O0",
    ];
    if (compileFlags) {
      const argv = compileFlags.split(" ");
      ascArgv = ascArgv.concat(argv);
    }
    const { error, stderr } = await main(ascArgv);
    if (error) {
      // eslint-disable-next-line @typescript-eslint/no-base-to-string
      console.error(stderr.toString());
      throw error;
    }
  };

  // Here, for-await is more efficient and less memory cost than Promise.all()
  for (const codePath of testCodePaths) {
    await compile(codePath);
  }

  return wasm;
}

function getNewPath(newFolder: string, oldFolder: string, srcPath: string): string {
  return join(newFolder, relative(oldFolder, srcPath)).replaceAll(/\\/g, "/");
}
