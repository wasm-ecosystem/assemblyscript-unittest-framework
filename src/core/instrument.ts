import initInstrumenter from "../../build_wasm/bin/wasm-instrumentation.js";
import { InstrumentResult } from "../interface.js";

export async function instrument(
  sourceWasms: string[],
  sourceCodePaths: string[],
  collectCoverage: boolean
): Promise<InstrumentResult[]> {
  const includeRegexs = sourceCodePaths.map((path) => {
    return `(start:)?${path.slice(0, -3)}.*`;
  });
  const includeFilter = JSON.stringify(includeRegexs);
  const res: InstrumentResult[] = [];
  const instrumenter = await initInstrumenter();
  for (const sourceFile of sourceWasms) {
    const baseName = sourceFile.slice(0, -5);
    const result = new InstrumentResult(baseName);

    const reportFunction = "__unittest_framework_env/traceExpression";

    const source = instrumenter.allocateUTF8(sourceFile);
    const output = instrumenter.allocateUTF8(result.instrumentedWasm);
    const report = instrumenter.allocateUTF8(reportFunction);
    const sourceMap = instrumenter.allocateUTF8(result.sourceMap);
    const debugInfo = instrumenter.allocateUTF8(result.debugInfo);
    const expectInfo = instrumenter.allocateUTF8(result.expectInfo);
    const include = instrumenter.allocateUTF8(includeFilter);

    instrumenter._wasm_instrument(
      source,
      output,
      report,
      sourceMap,
      expectInfo,
      debugInfo,
      include,
      0,
      true,
      collectCoverage
    );
    for (const ptr of [source, output, report, sourceMap, debugInfo, expectInfo, include]) {
      instrumenter._free(ptr);
    }
    res.push(result);
  }

  return res;
}
