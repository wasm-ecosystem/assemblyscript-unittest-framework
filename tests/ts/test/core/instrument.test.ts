import fs from "fs-extra";
import { join } from "node:path";
import { tmpdir } from "node:os";
import { fileURLToPath, URL } from "node:url";
import { compile } from "../../../../src/core/compile.js";
import { instrument } from "../../../../src/core/instrument.js";

const fixturePath = join(fileURLToPath(new URL(".", import.meta.url)), "..", "..", "fixture", "constructor.ts");
const outputDir = join(tmpdir(), "assemblyscript-unittest-framework");

test("Instrument", async () => {
  await compile([fixturePath], outputDir, "--memoryBase 16 --exportTable");
  const base = join(outputDir, "constructor");
  const wasmPath = join(outputDir, "constructor.wasm");
  const sourceCodePath = "tests/ts/fixture/constructor.ts";
  const results = await instrument([wasmPath], [sourceCodePath], true);
  expect(results.length).toEqual(1);
  const result = results[0]!;
  const instrumentedWasm = join(outputDir, "constructor.instrumented.wasm");
  const debugInfo = join(outputDir, "constructor.debugInfo.json");
  const expectInfo = join(outputDir, "constructor.expectInfo.json");
  expect(result.baseName).toEqual(base);
  expect(result.sourceWasm).toEqual(wasmPath);
  expect(result.instrumentedWasm).toEqual(instrumentedWasm);
  expect(result.debugInfo).toEqual(debugInfo);
  expect(result.expectInfo).toEqual(expectInfo);
  expect(fs.existsSync(instrumentedWasm)).toEqual(true);
  expect(fs.existsSync(debugInfo)).toEqual(true);
  expect(fs.existsSync(expectInfo)).toEqual(true);
  expect(fs.readFileSync(debugInfo, { encoding: "utf8" })).toMatchSnapshot();
});
