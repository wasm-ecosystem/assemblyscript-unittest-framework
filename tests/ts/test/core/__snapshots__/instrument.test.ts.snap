// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Instrument 1`] = `"{"debugFiles":["~lib/rt/common.ts","~lib/rt/tlsf.ts","~lib/shared/typeinfo.ts","~lib/rt/itcms.ts","tests/ts/fixture/constructor.ts","~lib/shared/runtime.ts","~lib/util/number.ts","~lib/util/math.ts","~lib/util/string.ts","~lib/rt.ts","~lib/util/error.ts"],"debugInfos":null}"`;
